### ISSUE TITLE
no way to support setuptools plugin option in setup.cfg metadata

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [pypa/setuptools#1055]
while experimenting with setuptools_scm metadata i discovered that there doesn't seem to be a way to add setup.cfg metadata support to setuptools plugins

**Comment by pganssle** [comment]
I think I generally get what you mean, but @RonnyPfannschmidt can you provide a minimal example of what you have in mind?

**Comment by pganssle** [comment]
@RonnyPfannschmidt Well, plugins can put their configuration wherever they want, but there's standard library support for `setup.cfg`, but not for `pyproject.toml`, and you're asking for more stuff to go into `setup.cfg` in the first place...

In any case, I'm just saying there's already a well-established place for tools to put their configuration in `setup.cfg`, and usually when options are added to `setup.cfg`, they go under a heading *not* present in `setup.py` - note that `setup()` doesn't look like this:

```python
setup(metadata=dict(
    name='mypkg'
    version='0.0.1'))
```

The fact that `setup()` takes a bunch of loose keys for the configuration doesn't mean that `setup.cfg` can't expect those keys to be namespaced or categorized. To me the natural place to namespace such keys would be under tool-specific configuration headers.

Now, I don't know how `pytest` implements this `[tool:pytest]` parser. If it's super inconvenient it makes sense for `setuptools` to provide some hooks for it, but I think that that's the rough shape that the solution should take, and it's *already available today*.

**Comment by jaraco** [comment]
What setuptools_scm needs is not a place to define configuration, but a way for declarative config to trigger the hook that would import setuptools_scm and allow it to customize the Distribution.

**Comment by pganssle** [comment]
@jaraco Ah, these things are registered at install-time? For some reason I thought you manually imported your plugins in the install script.

That changes things pretty significantly, and I'm much closer to agreeing that this should be done. Maybe something like this:

```ini
[setup_plugins]
use_scm_version=True
```

I'm pretty much 90% sold on this. The only reservation I have is that it makes it so that the "declarative" configuration file is basically invoking arbitrary scripts, which sorta moves the needle away from it being declarative. Still, I think there's a case to be made that, used sanely, this is no different from having more features in `setuptools` itself, but allows for those features to be broken out into extensions to allow those features to be "opt in".

**Comment by con-f-use** [comment]
Two questions regarding this:
 1.) How likely is it to be permanent now. In my company, we would like to write a module that makes heavy use of this functionality, but I won't if it's likely to be changed much or taken out in the future.
 2.) How do I actually set a priority for a `finalize_options` entrypoint?
   ```
   [setuptools.finalize_distribution_options]
   testfin = mypackage.integration:finalize
   ```
   and finalize is a function. I tried this:
   ```
   # In mypackage/integration.py
   def finalize(dist):
       # do something with dist
   finalize.order = 99
   ```
   but the order is still `0`.

**Comment by jaraco** [comment]
>     1. How likely is it to be permanent now. In my company, we would like to write a module that makes heavy use of this functionality, but I won't if it's likely to be changed much or taken out in the future.

I'd say unlikely. If you're motivated to lock in this behavior, I'd accept docs or tests updates to establish the behavior more officially.



>     2\. How do I actually set a priority for a `finalize_options` entrypoint?

This behavior is broken. Would you open a new issue? As I review it now, I see that the sort happens before the callable is resolved. So that bug will need to be fixed before you can rely on it.

I agree, your patch would likely address the issue, though it causes `load()` to be called twice. I'd recommend another approach that calls `load()` just once.

I would omit the `int()` cast. Any value that's comparable to others should be suitable (including `float('inf')`, `-float('inf')`, other floats, or even a custom object that can compare as greater or equal to numeric values.

> Btw. if it's an issue, it would not invalidate @jaraco random suffle test from #1877 as that did not rely on the `order` attribute, but suffelled directly.

I don't see any tests in that PR, but I'd welcome them.

**Comment by con-f-use** [comment]
Okay, I will open an issue and then very likely do a PR to fix it tomorrow.

------

I didn't mean test in the pytest / software testing kind of way. I was referring this bit:
```python
        eps = list(pkg_resources.iter_entry_points(hook_key))
        random.shuffle(eps)
        for ep in eps:
```
But never mind, not that important.

**Comment by Thebaldguy15** [comment]
> @pganssle note that pytests `[tool:pytest]` is fundamentally broken and scheduled for removal as the default parser for `setup.cfg` and pytests own config parser fundamentally differ
> 
> i'm asking for a mechanism similar to the `distutils.setup_keywords` mechanism - it should be some kind of `distutils.config_section`, `setuptools.metadata.key` and `setuptools.options.key` to allow extensions to get their own data in shape

**Comment by jaraco** [comment]
@Thebaldguy15 By quoting that paragraph, are you attempting to revive that conversation.

Since we had this conversation, it's my understanding that the best recommendation has shifted to asking each plugin to resolve their own configuration as addressed in #2155. If you have further considerations that aren't addressed, please raise them in a new issue.
